<launch>
  <arg name="ns" default="perception"/>
  <arg name="input_image" default="/kinect_head/rgb/image_rect_color"/>
  <arg name="reset_track" default="/cutie_node/reset"/>
  <arg name="cutie_prompt" default="/cutie_node/process_prompt"/>
  <arg name="vqa" default="false"/>
  <arg name="device" default="cuda:0"/>

  <arg name="tracking" default="true"/>
  <arg name="dino_node" default="/grounding_dino_node"/>
  <arg name="yolo_node" default="/yolo_node"/>
  <arg name="vlpart_node" default="/vlpart_node"/>
  <arg name="clip_node" default="/clip_node"/>
  <arg name="vqa_node" default="/vqa_node"/>
  <arg name="cutie_node" default="/cutie_node"/>
  <arg name="sam_node" default="/sam_node"/>

  <arg name="yolo_model" default="yolov8x-worldv2.pt" /> <!-- YOLO model, refer ultralytics yolo world model -->
  <arg name="get_mask" default="false" /> <!-- get mask when grounding detection -->
  <arg name="with_bbox" default="true" /> <!-- false if faster, but for visualization -->
  <arg name="refine_mask" default="true" /> <!-- refine mask predicts twice to refine mask -->

  <!-- vlpart args -->
  <arg name="vlpart_node_name" default="vlpart_node"/>
  <arg name="vocabulary" default="custom" /> <!-- "lvis", "paco"... "custom" -->
  <arg name="classes" default="cup handle; bottle cap;" /> <!-- object to detect when vocabulary is custom -->
  <arg name="confidence_threshold" default="0.5" /> <!-- confidence threshold -->
  <arg name="use_sam" default="false" /> <!-- refine mask predicts twice to refine mask -->
  <arg name="track" default="false" />

  <!-- sam args -->
  <arg name="model_type" default="vit_t" /> <!-- SAM model : vit_t, vit_b, vit_l, vit_h ,vit_b_hq, vit_l_hq, vit_h_hq -->
  <arg name="mode" default="prompt" /> <!-- SAM mode : prompt, automatic -->
  <arg name="num_slots" default="10" /> <!-- max num of mask generation when automatic mode -->
  <arg name="area_threshold" default="400" /> <!-- refine mask area threshold -->
  <arg name="refine_mode" default="holes" /> <!-- refine mode : "holes", "islands" -->

  <!-- tracker args -->

  <group ns="$(arg ns)">
    <node name="perception_reconfigure_node" pkg="llm_ros" type="perception_reconfigure_node.py" output="screen" >
      <remap from="~input_image" to="$(arg input_image)" />
      <remap from="~input_segmentation" to="$(arg input_image)" />
      <remap from="~reset_prompt" to="$(arg cutie_prompt)" />
      <rosparam subst_value="true" >
          tracking: $(arg tracking)
          dino_node: $(arg dino_node)
          yolo_node: $(arg yolo_node)
          vlpart_node: $(arg vlpart_node)
          clip_node: $(arg clip_node)
          vqa_node: $(arg vqa_node)
          cutie_node: $(arg cutie_node)
          sam_node: $(arg sam_node)
      </rosparam>
    </node>
  </group>

  <!-- <!-\- perception nodes -\-> -->
  <!-- <node name="clip_inference_node" pkg="llm_ros" type="clip_inference_node.py" output="screen" > -->
  <!--   <remap from="~input_image" to="$(arg input_image)" /> -->
  <!--   <rosparam subst_value="true" > -->
  <!--       device: $(arg device) -->
  <!--   </rosparam> -->
  <!-- </node> -->

  <!-- <node name="grounding_dino_node" pkg="tracking_ros" type="grounding_dino_node.py" output="screen" > -->
  <!--   <remap from="~input_image" to="$(arg input_image)" /> -->
  <!--   <rosparam subst_value="true" > -->
  <!--       device: $(arg device) -->
  <!--       get_mask: $(arg get_mask) -->
  <!--       model_type: $(arg model_type) -->
  <!--       refine_mask: $(arg refine_mask) -->
  <!--   </rosparam> -->
  <!-- </node> -->

  <!-- <node name="yolo_node" pkg="tracking_ros" type="yolo_node.py" output="screen" > -->
  <!--   <remap from="~input_image" to="$(arg input_image)" /> -->
  <!--   <remap from="~output/segmentation" to="$(arg input_segmentation)" /> -->
  <!--   <rosparam subst_value="true" > -->
  <!--       device: $(arg device) -->
  <!--       get_mask: $(arg get_mask) -->
  <!--       model_id: $(arg yolo_model) -->
  <!--       model_type: $(arg model_type) -->
  <!--       refine_mask: $(arg refine_mask) -->
  <!--   </rosparam> -->
  <!-- </node> -->

  <!-- <node name="vlpart_reconfigure_node" -->
  <!--         pkg="tracking_ros" type="vlpart_reconfigure_node.py" -->
  <!--         output="screen" if="$(arg reconfigure)" > -->
  <!--   <rosparam subst_value="true" > -->
  <!--       input_image: $(arg input_image) -->
  <!--       node_name: /$(arg vlpart_node_name) -->
  <!--       device: $(arg device) -->
  <!--       use_sam: $(arg use_sam) -->
  <!--       vocabulary: $(arg vocabulary) -->
  <!--       classes: $(arg classes) -->
  <!--       confidence_threshold: $(arg confidence_threshold) -->
  <!--   </rosparam> -->
  <!-- </node> -->

  <!-- <!-\- vlpart node -\-> -->
  <!-- <node name="$(arg vlpart_node_name)" -->
  <!--         pkg="tracking_ros" type="vlpart_node.py" -->
  <!--         output="screen" unless="$(arg reconfigure)" > -->
  <!--   <remap from="~input_image" to="$(arg input_image)" /> -->
  <!--   <rosparam subst_value="true" > -->
  <!--       device: $(arg device) -->
  <!--       use_sam: $(arg use_sam) -->
  <!--       vocabulary: $(arg vocabulary) -->
  <!--       classes: $(arg classes) -->
  <!--       confidence_threshold: $(arg confidence_threshold) -->
  <!--   </rosparam> -->
  <!-- </node> -->

  <!-- <!-\- sam node -\-> -->
  <!-- <node name="sam_node" pkg="tracking_ros" type="sam_node.py" output="screen" if="$(arg use_sam)"> -->
  <!--   <remap from="~input_image" to="$(arg input_image)" /> -->
  <!--   <rosparam subst_value="true" > -->
  <!--       device: $(arg device) -->
  <!--       model_type: $(arg model_type) -->
  <!--       mode: $(arg mode) -->
  <!--       num_slots: $(arg num_slots) -->
  <!--       interactive_mode: false -->
  <!--       refine_mask: $(arg refine_mask) -->
  <!--   </rosparam> -->
  <!-- </node> -->

  <node name="cutie_node" pkg="tracking_ros" type="cutie_node.py" output="screen" if="$(arg tracking)">
    <remap from="~input_image" to="$(arg input_image)" />
    <!-- <remap from="~input_segmentation" to="$(arg input_segmentation)" /> -->
    <remap from="~process_prompt" to="$(arg cutie_prompt)" />
    <remap from="~reset" to="$(arg reset_track)" />
    <rosparam subst_value="true" >
        device: $(arg device)
        with_bbox: $(arg with_bbox)
    </rosparam>
  </node>

  <!-- <include file="$(find tracking_ros)/launch/grounding_detection.launch" > -->
  <!--   <arg name="get_mask" value="true"/> -->
  <!--   <arg name="track" value="true"/> -->
  <!-- </include> -->
  <!-- <include file="$(find tracking_ros)/launch/tracking_3d.launch" /> -->
  <!-- <include file="$(find vlm_ros)/launch/sample_vqa.launch" if="$(arg vqa)"> -->
  <!--   <arg name="input_image" value="$(arg input_image)"/> -->
  <!-- </include> -->
</launch>
